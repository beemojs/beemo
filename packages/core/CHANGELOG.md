# 1.0.0

#### üí• Breaking

- Migrated to the new [@boost/event](https://milesj.gitbook.io/boost/event) system.
- Script names (on the command line) will now error if not in kebab case.
- Scripts and Drivers now require an explicit `blueprint` method.
- Renamed `--live` option to `--stdio`.
- Renamed `--priority` option to `--graph`.
- Renamed `execute.priority` setting to `execute.graph`.
- Renamed `Context#root` to `cwd`.
- Renamed `Driver#handleFailure` to `processFailure`.
- Renamed `Driver#handleSuccess` to `processSuccess`.
- Removed `DriverContext#eventName` and `ScriptContext#eventName`.
- Removed `ScriptContext#binName`. Use `scriptName` instead.

#### üöÄ Updates

- Added a new package,
  [@beemo/dependency-graph](https://www.npmjs.com/package/@beemo/dependency-graph), to handle the
  dependency resolution.
- Added a `none` strategy to `Driver`s. With this strategy, the consumer will need to manually
  create a config file.
- Added a `versionOption` metadata setting to `Driver`s.
- Added `Script#executeCommand`, so that local binaries can easily be executed.
- Added optional `name` support to scaffolding.
- Updated the driver/script execution pipeline to process in parallel batches.

#### üêû Fixes

- Fixed script `this` scope being lost within script tasks.
- Fixed the Beemo emoji not appearing in the console.

#### üõ† Internals

- `Beemo` now extends from `Tool` instead of managing an instance.
- Updated `hygen` to v4.
- Updated dependencies.
