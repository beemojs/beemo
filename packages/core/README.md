# ðŸ¤– Beemo

[![Build Status](https://github.com/beemojs/beemo/workflows/Build/badge.svg)](https://github.com/beemojs/beemo/actions?query=branch%3Amaster)
[![npm version](https://badge.fury.io/js/%40beemo%2Fcore.svg)](https://www.npmjs.com/package/@beemo/core)
[![npm deps](https://david-dm.org/beemojs/beemo.svg?path=packages/core)](https://www.npmjs.com/package/@beemo/core)

Manage dev and build tools, their configuration, and commands in a single centralized repository.
Beemo aims to solve the multi-project maintenance fatigue by removing the following burdens across
all projects: config and dotfile management, multiple config patterns, up-to-date development
dependencies, continuous copy and paste, and more.

### Features

- Manage dev tools and configurations in a single repository.
- Configure supported dev tools using `.js` files.
- Customize and alter config at runtime with CLI options.
- Pass custom CLI options to dev tool commands without failure.
- Automatically expand glob patterns (a better alternative to bash).
- Listen to and act upon events.
- Easily share config between dev tools.
- Avoid relative config or `extends` paths.
- Automatic config file cleanup.
- Custom scripts with CLI options.
- Scaffolding and template generation.
- Workspaces (monorepo) support.
- Automated TypeScript project references.
- Parallel, pooled, and prioritized builds.
- Integrated watch/live mode.
- And much more.

## Requirements

- Node 8.9+
- GitHub, Bitbucket, or another VCS

## Documentation

[https://milesj.gitbook.io/beemo](https://milesj.gitbook.io/beemo)
