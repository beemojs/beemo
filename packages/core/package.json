{
  "name": "@beemo/core",
  "version": "2.0.0-alpha.0",
  "description": "Manage all dev tool configurations in a single centralized repository.",
  "keywords": [
    "beemo",
    "build",
    "tool",
    "config",
    "manager",
    "babel",
    "eslint",
    "flow",
    "jest",
    "prettier",
    "typescript"
  ],
  "main": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "engines": {
    "node": ">=10.17.0"
  },
  "repository": "https://github.com/beemojs/beemo/tree/master/packages/core",
  "author": "Miles Johnson",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "dependencies": {
    "@beemo/dependency-graph": "^2.0.0-alpha.0",
    "@boost/args": "^2.3.1",
    "@boost/common": "^2.5.1",
    "@boost/config": "^2.3.1",
    "@boost/debug": "^2.2.1",
    "@boost/event": "^2.3.1",
    "@boost/pipeline": "^2.2.1",
    "@boost/plugin": "^2.2.1",
    "@boost/terminal": "^2.2.0",
    "@boost/translate": "^2.2.1",
    "chalk": "^4.1.0",
    "execa": "^5.0.0",
    "fast-glob": "^3.2.5",
    "fs-extra": "^9.1.0",
    "hygen": "^6.0.4",
    "is-glob": "^4.0.1",
    "lodash": "^4.17.15",
    "micromatch": "^4.0.2"
  },
  "devDependencies": {
    "@boost/test-utils": "^2.2.0",
    "@types/micromatch": "^4.0.1"
  }
}
