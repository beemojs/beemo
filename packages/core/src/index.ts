/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Blueprint, PackageStructure, Path, PortablePath, Predicates } from '@boost/common';

export * from './constants';
export * from './contexts/ConfigContext';
export * from './contexts/Context';
export * from './contexts/DriverContext';
export * from './contexts/ScaffoldContext';
export * from './contexts/ScriptContext';
export * from './Driver';
export * from './Script';
export * from './Tool';
export * from './types';

export { Path };
export type { Blueprint, PackageStructure, PortablePath, Predicates };
