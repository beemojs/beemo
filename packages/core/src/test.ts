import execa from 'execa';
import { ArgList, Arguments, Argv } from '@boost/args';
import { Path } from '@boost/common';
import { mockDebugger } from '@boost/debug/test';
import {
	ConfigContext,
	ConfigFile,
	Context,
	Driver,
	DriverContext,
	DriverContextOptions,
	DriverMetadata,
	ScaffoldContext,
	ScaffoldContextOptions,
	Script,
	ScriptContext,
	ScriptContextOptions,
	Tool,
} from '.';

export { mockDebugger };

const TEST_ROOT = new Path(process.env.BEEMO_TEST_ROOT ?? process.cwd());

export class TestDriver<O extends object = {}> extends Driver<O> {
	override readonly name = 'test-driver';
}

export class TestScript<O extends object = {}> extends Script<O> {
	override readonly name = 'test-script';

	execute(): any {
		return Promise.resolve();
	}
}

export function mockConsole<K extends keyof Console>(name: K): jest.SpyInstance {
	return jest.spyOn(console, name as 'log').mockImplementation(() => {});
}

export function mockToolConfig(): ConfigFile {
	return {
		configure: {
			cleanup: false,
			parallel: true,
		},
		debug: false,
		drivers: [],
		execute: {
			concurrency: 1,
			graph: true,
			output: '',
		},
		module: '@local',
		scripts: [],
		settings: {},
	};
}

export function mockTool(argv: Argv = []): Tool {
	const tool = new Tool({
		argv,
		cwd: TEST_ROOT,
	});

	// @ts-expect-error Allow readonly
	tool.debug = mockDebugger();

	tool.config = mockToolConfig();

	return tool;
}

export function mockDriver<C extends object = {}>(
	name: string,
	tool: Tool | null = null,
	metadata: Partial<DriverMetadata> = {},
): Driver<C> {
	const driver = new TestDriver<C>();

	// @ts-expect-error For testing purposes
	driver.name = name;
	driver.tool = tool ?? mockTool();

	driver.setMetadata({
		bin: name.toLowerCase(),
		configName: `${name}.json`,
		title: name,
		...metadata,
	});

	return driver;
}

export function mockScript(name: string, tool: Tool | null = null): Script<{}> {
	const script = new TestScript<{}>();

	// @ts-expect-error For testing purposes
	script.name = name;
	script.tool = tool ?? mockTool();

	return script;
}

export function stubArgs<T extends object>(options: T, params: ArgList = []): Arguments<T> {
	return {
		command: [],
		errors: [],
		options,
		params,
		rest: [],
		unknown: {},
	};
}

export function stubConfigArgs(): Arguments<{}> {
	return stubArgs({});
}

export function stubDriverArgs(
	fields?: Partial<DriverContextOptions>,
): Arguments<DriverContextOptions> {
	return stubArgs({
		concurrency: 1,
		graph: false,
		workspaces: '',
		...fields,
	});
}

export function stubScaffoldArgs(
	fields?: Partial<ScaffoldContextOptions>,
): Arguments<ScaffoldContextOptions> {
	return stubArgs({
		dry: false,
		...fields,
	});
}

export function stubScriptArgs(
	fields?: Partial<ScriptContextOptions>,
): Arguments<ScriptContextOptions> {
	return stubArgs({
		concurrency: 1,
		graph: false,
		workspaces: '',
		...fields,
	});
}

export function applyContext<T extends Context>(context: T): T {
	context.args = stubArgs({ a: true, foo: 'bar' }, ['baz']);
	context.argv = ['-a', '--foo', 'bar', 'baz'];
	context.cwd = TEST_ROOT;
	context.configModuleRoot = TEST_ROOT;
	context.workspaceRoot = TEST_ROOT;
	context.workspaces = [];

	return context;
}

export function stubConfigContext(): ConfigContext {
	return applyContext(new ConfigContext(stubArgs({})));
}

export function stubDriverContext(driver?: Driver): DriverContext {
	return applyContext(new DriverContext(stubDriverArgs(), driver ?? new TestDriver()));
}

export function stubScriptContext(script?: Script): ScriptContext {
	const context = applyContext(new ScriptContext(stubScriptArgs(), 'script'));

	if (script) {
		context.setScript(script);
	}

	return context;
}

export function stubScaffoldContext(
	generator: string = 'generator',
	action: string = 'action',
	name: string = '',
): ScaffoldContext {
	return applyContext(new ScaffoldContext(stubScaffoldArgs(), generator, action, name));
}

export function prependRoot(part: string): Path {
	return TEST_ROOT.append(part);
}

export function getRoot(): Path {
	return TEST_ROOT;
}

export function stubExecResult(fields?: Partial<execa.ExecaReturnValue>): execa.ExecaReturnValue {
	return {
		all: '',
		command: '',
		escapedCommand: '',
		exitCode: 0,
		failed: false,
		isCanceled: false,
		killed: false,
		signal: undefined,
		stderr: '',
		stdout: '',
		timedOut: false,
		...fields,
	};
}
