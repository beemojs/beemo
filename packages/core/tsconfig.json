{
  "compilerOptions": {
    "declarationDir": "lib",
    "outDir": "lib",
    "rootDir": "src"
  },
  "exclude": [
    "lib",
    "tests"
  ],
  "extends": "../../tsconfig.options.json",
  "include": [
    "src/**/*",
    "types/**/*",
    "../../types/**/*"
  ],
  "references": []
}