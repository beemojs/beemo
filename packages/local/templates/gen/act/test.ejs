---
to: TEST.md
---
This simply verifies that scaffolding is working correctly.
