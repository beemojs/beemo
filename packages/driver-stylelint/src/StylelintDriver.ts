import fs from 'fs';
import { ConfigContext, Driver, Path } from '@beemo/core';
import { Event } from '@boost/event';
import { StylelintArgs, StylelintConfig } from './types';

// Success: Writes warnings to stdout
// Failure: Writes failures to stdout
export default class StylelintDriver extends Driver<StylelintConfig> {
  readonly name = '@beemo/driver-stylelint';

  readonly onCreateIgnoreFile = new Event<[ConfigContext, Path, { ignore: string[] }]>(
    'create-ignore-file',
  );

  bootstrap() {
    this.setMetadata({
      bin: 'stylelint',
      configName: '.stylelintrc.js',
      description: this.tool.msg('app:stylelintDescription'),
      title: 'stylelint',
    });

    this.onCreateConfigFile.listen(this.handleCreateIgnoreFile);
  }

  /**
   * If an "ignore" property exists in the stylelint config, create a ".sylelintignore" file.
   */
  private handleCreateIgnoreFile = (
    context: ConfigContext<StylelintArgs>,
    configPath: Path,
    config: StylelintConfig,
  ) => {
    if (!config.ignore) {
      return;
    }

    if (!Array.isArray(config.ignore)) {
      throw new TypeError(this.tool.msg('errors:stylelintIgnoreInvalid'));
    }

    const ignorePath = configPath.parent().append('.stylelintignore');
    const { ignore } = config;

    this.onCreateIgnoreFile.emit([context, ignorePath, { ignore }]);

    fs.writeFileSync(ignorePath.path(), ignore.join('\n'));

    // Add to context so that it can be automatically cleaned up
    context.addConfigPath('stylelint', ignorePath);

    delete config.ignore;
  };
}
