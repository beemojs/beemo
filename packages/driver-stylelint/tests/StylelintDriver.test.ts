import fs from 'fs';
import { DriverContext, Path } from '@beemo/core';
import { mockTool, stubDriverContext, stubExecResult } from '@beemo/core/test';
import StylelintDriver from '../src/StylelintDriver';

describe('StylelintDriver', () => {
  let driver: StylelintDriver;
  let context: DriverContext;
  let writeSpy: jest.SpyInstance;

  beforeEach(() => {
    driver = new StylelintDriver();
    driver.tool = mockTool();
    driver.bootstrap();

    context = stubDriverContext(driver);

    writeSpy = jest.spyOn(fs, 'writeFileSync').mockImplementation(() => true);
  });

  afterEach(() => {
    writeSpy.mockRestore();
  });

  it('sets options from constructor', () => {
    driver = new StylelintDriver({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: 'true' },
    });

    expect(driver.options).toEqual({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: 'true' },
      expandGlobs: true,
      strategy: 'native',
    });
  });

  it('sets correct metadata', () => {
    expect(driver.metadata).toEqual(
      expect.objectContaining({
        bin: 'stylelint',
        configName: '.stylelintrc.js',
        configOption: '--config',
        dependencies: [],
        description: 'Lint styles with stylelint',
        filterOptions: true,
        helpOption: '--help',
        title: 'stylelint',
        useConfigOption: false,
      }),
    );
  });

  describe('mergeConfig()', () => {
    it('merges arrays by unique value', () => {
      expect(
        driver.mergeConfig(
          {
            extends: 'abc',
            rules: {
              foo: 'error',
            },
          },
          {
            rules: {
              foo: ['error', 'always'],
            },
          },
        ),
      ).toEqual({
        extends: 'abc',
        rules: {
          foo: ['error', 'always'],
        },
      });
    });

    it('merges ignore list correctly', () => {
      expect(
        driver.mergeConfig(
          {
            ignore: ['foo', 'bar'],
          },
          {
            ignore: ['baz', 'foo'],
          },
        ),
      ).toEqual({
        ignore: ['foo', 'bar', 'baz'],
      });
    });
  });

  describe('processFailure()', () => {
    it('outputs stderr and stdout', () => {
      driver.processFailure(
        stubExecResult({
          command: 'stylelint',
          stderr: 'Error',
        }),
      );

      expect(driver.output.stderr).toBe('Error');
    });
  });

  describe('handleCreateIgnoreFile()', () => {
    it('does nothing if no ignore field', () => {
      const config = { extends: 'abc' };

      driver.onCreateConfigFile.emit([context, new Path('/some/path/.stylelintrc.js'), config]);

      expect(config).toEqual({ extends: 'abc' });
    });

    it('errors if not an array or string', () => {
      expect(() => {
        driver.onCreateConfigFile.emit([
          context,
          new Path('/some/path/.stylelintrc.js'),
          {
            // @ts-expect-error
            ignore: 'abc',
          },
        ]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('creates ignore file and updates references', () => {
      const config = {
        extends: 'abc',
        ignore: ['foo', 'bar', 'baz'],
      };

      driver.onCreateConfigFile.emit([context, new Path('/some/path/.stylelintrc.js'), config]);

      expect(writeSpy).toHaveBeenCalledWith('/some/path/.stylelintignore', 'foo\nbar\nbaz');

      expect(context.configPaths).toEqual([
        { driver: 'stylelint', path: new Path('/some/path/.stylelintignore') },
      ]);

      expect(config).toEqual({ extends: 'abc' });
    });

    it('emits `onCreateIgnoreFile` event', () => {
      const createSpy = jest.fn((ctx, path, config) => {
        config.ignore.push('qux');
      });

      driver.onCreateIgnoreFile.listen(createSpy);

      const config = {
        extends: 'abc',
        ignore: ['foo', 'bar', 'baz'],
      };

      driver.onCreateConfigFile.emit([context, new Path('/some/path/.stylelintrc.js'), config]);

      expect(createSpy).toHaveBeenCalledWith(context, new Path('/some/path/.stylelintignore'), {
        ignore: ['foo', 'bar', 'baz', 'qux'],
      });

      expect(writeSpy).toHaveBeenCalledWith('/some/path/.stylelintignore', 'foo\nbar\nbaz\nqux');
    });
  });
});
