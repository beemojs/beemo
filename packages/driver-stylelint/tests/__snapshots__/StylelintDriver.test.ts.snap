// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StylelintDriver handleCreateIgnoreFile() errors if not an array or string 1`] = `"Ignore configuration must be an array of strings."`;
